nexus/reconfigurator/planning/src/blueprint_builder/external_networking.rs

```diff
@@ -149,6 +171,17 @@ impl<'a> BuilderExternalNetworking<'a> {
             }
         }
 
+        // Recycle the IP addresses of expunged external DNS zones.
+        // TODO: Remove when external DNS addresses come from policy.
+        let mut available_external_dns_ips: BTreeSet<IpAddr> = BTreeSet::new();
+        for (_, zone) in
+            parent_blueprint.all_omicron_zones(BlueprintZoneFilter::Expunged)
+        {
+            if let BlueprintZoneType::ExternalDns(dns) = &zone.zone_type {
+                available_external_dns_ips.insert(dns.dns_address.addr.ip());
```

Comment:
Do we need to guard this with an "...and also this IP is not used by a non-expunged zone"? The specific sequence I'm thinking about is:

* We have 3 in-service zones with IPs A,B,C
* We expunge the zone with IP A
* We do a planning run; we add A to this set, and start a new in-service zone; we now have a blueprint with an expunged zone with IP A, and in-service zones with IPs A,B,C
* We do another planning run; we see an expunged zone with IP A, so we add it here, even though we already replaced it with a new in-service zone

Reply:
An excellent idea; strictly safer, no downsides that I can see. Implemented in `2a4228648`.